--- nvidia-current-313.18/NVIDIA-Linux-x86_64-313.18/kernel/conftest.sh~	2013-02-24 14:48:37.458218751 +0100
+++ nvidia-current-313.18/NVIDIA-Linux-x86_64-313.18/kernel/conftest.sh	2013-02-24 14:48:41.168269125 +0100
@@ -1696,15 +1696,9 @@ case "$6" in
                 # a Makefile and a .config file. If this is a 2.6
                 # kernel older than 2.6.6, that's all we require to
                 # build the module.
-                #
-                PATCHLEVEL=$(grep "^PATCHLEVEL =" $MAKEFILE | cut -d " " -f 3)
-                SUBLEVEL=$(grep "^SUBLEVEL =" $MAKEFILE | cut -d " " -f 3)
 
-                if [ -n "$PATCHLEVEL" -a $PATCHLEVEL -ge 6 \
-                        -a -n "$SUBLEVEL" -a $SUBLEVEL -le 5 ]; then
-                    SELECTED_MAKEFILE=Makefile.kbuild
-                    RET=0
-                fi
+                SELECTED_MAKEFILE=Makefile.kbuild
+                RET=0
             fi
         fi
 
